Add allowed iframe host setting

alxnddr opened the pull request:
Closes https://github.com/metabase/metabase-private/issues/238

### Description

Adds a setting to configure allowed iframe hosts under Admin -> General.

<img width="721" alt="Screenshot 2024-10-22 at 8 25 22 PM" src="https://github.com/user-attachments/assets/f16f3f8b-9358-4f75-b435-fab17ed28425">

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

alxnddr committed:
add allowed iframe host setting wip

adam-james-v committed:
use allowed-iframe-hosts setting in the CSP header

adam-james-v committed:
add a test for the frame-src csp directive

kulyk committed:
Update allowed-iframe-hosts setting definition

kulyk committed:
Add error state for forbidden iframe url domains

kulyk committed:
Move out iframe e2e test suite

kulyk committed:
Add e2e test

kulyk committed:
Update error message in view mode

kulyk committed:
Merge branch 'master' into add-allowed-iframe-hosts-setting

kulyk committed:
Fix unit tests

kulyk committed:
Update setting on the admin page

kulyk committed:
Update error state

kulyk committed:
Add links to error states

kulyk committed:
Update docs links

kulyk committed:
Update link anchors

adam-james-v committed:
Add default allowed hosts to public setting

kulyk committed:
Update allowed domain check logic

kulyk committed:
Fix default value display in admin page

kulyk committed:
Don't update setting without changes

kulyk committed:
Update error message spacing

alxnddr commented on src/metabase/public_settings.clj:
```diff
@@ -175,6 +175,36 @@
   :export?    true)
 
+(def ^:private default-allowed-iframe-hosts
+  "youtube.com,
```
Let's add `youtu.be` as well

adam-james-v committed:
correct the parsing of allowed-hosts string for CSP header entries

adam-james-v committed:
fix test

alxnddr committed:
fix not handling wildcard ports

dpsutton commented:
Couple things we will need to address:
1. This gets called a _lot_. Instrumented it and reloaded on a question invoked this 20 times. This doesn't do a tiny amount of work: we are parsing lots of strings, detecting hosts, etc. It's non-substantial to do on every request. It's also deterministic on its input. We should memoize it in some manner.
2. I'm not sure if you are able to restore the current behavior of `*`: `security=> (parse-allowed-iframe-hosts "*") ["*:*"]`
3. The parser is weird about trailing slashes and a provided www. It's a bit of a rough edge:
```
security=> (parse-allowed-iframe-hosts "www.google.com")
["www.google.com"]
security=> (parse-allowed-iframe-hosts "google.com")
["google.com" "*.google.com"]
security=> (parse-allowed-iframe-hosts "google.com/")
[]
security=> (parse-allowed-iframe-hosts "docs.google.com")
["docs.google.com"]
```
I don't know what the right answers are here, but i don't see it discussed anywhere.
4. We use `log/error` here. I'm not sure what the level should be here. I'm not sure what else `parse-url` is used for that it is erroneous. Is an invalid url here erroneous? How noisy should that be?
5. `parse-allowed-iframe-hosts` returns either a vector of hosts `["google.com" "*.google.com"]` or a string error message. This is weird. It should _Log_ an error message and not return the `(catch Exception e (ex-message e))` instead of the parsed hosts.

kulyk committed:
Fix failing e2e test

kulyk committed:
Fix subdomain test

kulyk committed:
Merge branch 'master' into add-allowed-iframe-hosts-setting

adam-james-v committed:
address review

 - the parse-allowed-iframe-hosts fn is now memoized
 - a * entry is handled and doesn't produce a weird *:* entry
 - no more try/catch, errors in parsing will be logged but the list returns all valid entries
 - when www. is encountered, an entry including www. is added
 - trailing / is 'cleaned' and the entry is used as if there was no trailing /

dpsutton committed:
Fixup test for expecting a few more frame sources

dpsutton committed:
indentation fix for linter 🥲

kulyk committed:
Fix type error
